### PR TITLE
[WIP] Version 0.4

### PR DESCRIPTION
**Pull request by TOTBWF**
This release of refinery has a bunch of cool new features, and code cleanup.

The big one is the introduction of the concept of an "unsolvable hole", which generalizes the idea of errors found in previous versions of the library. This lets us continue on in the face of failures if we so choose, which is very useful for good error reporting.

# Checklist
- [ ] When `Commit` fails on both branches, don't discard the results from the LHS
- [ ] `reify` and `resume` should have error aware variants

### CODE REVIEW

File: src/Refinery/ProofState.hs [hunk 1]
```diff
@@ -36,19 +35,42 @@ import           Control.Monad.Logic
...
 
+-- | The core data type of the library.
+--
+-- NOTE: We need to split up the extract type into @ext'@ and @ext@, as
+-- we want to be functorial (and monadic) in @ext@, but it shows up in both
+-- co and contravariant positions. Splitting the type variable up into two solves this problem,
+-- at the cost of looking ugly.
 data ProofStateT ext' ext err s m goal
     = Subgoal goal (ext' -> ProofStateT ext' ext err s m goal)
+    -- ^ Represents a subgoal, and a continuation that tells
```
Comment (isovector) [inline src/Refinery/ProofState.hs:68]: :heart_eyes: for documentation

File: src/Refinery/ProofState.hs [hunk 2]
```diff
@@ -36,19 +35,42 @@ import           Control.Monad.Logic
...
+    -- This is useful if the first argument produces potentially infinite results.
     | Commit (ProofStateT ext' ext err s m goal) (ProofStateT ext' ext err s m goal)
+    -- ^ 'Commit' runs the first proofstate, and only runs the second if the first
+    -- does not produce any successful results.
     | Empty
-    | Failure err
+    -- ^ Silent failure. Always causes a backtrack, unlike 'Failure'.
+    | Failure err (ext' -> ProofStateT ext' ext err s m goal)
+    -- ^ This does double duty, depending on whether or not the user calls 'proofs'
+    -- or 'partialProofs'. In the first case, we ignore the continutation, and backtrack.
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: Do the errors still show up in `proofs`?

File: src/Refinery/ProofState.hs [hunk 3]
```diff
@@ -155,8 +181,49 @@ proofs s p = go s [] p
...
+    | PartialProof ext s [goal] [err]
+    -- ^ A so called "partial proof". These are proofs that encountered errors
+    -- during execution.
+    deriving (Eq, Show, Generic)
+
+isSuccessful :: PartialProof ext s goal err -> Bool
+isSuccessful SuccessfulProof {} = True
+isSuccessful PartialProof {} = False
+
+partialProofs :: forall ext err s m goal. (MonadExtract ext m) => s -> ProofStateT ext ext err s m goal -> m [PartialProof ext s goal err]
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: haddock + commentary for this would be greatly apprecaited

File: src/Refinery/Tactic.hs [hunk 4]
```diff
@@ -125,18 +135,44 @@ ensure p f t = check >> t
...
+-- | @tweak f t@ lets us modify the extract produced by the tactic @t@.
+tweak :: (Functor m) => (ext -> ext) -> TacticT jdg ext err s m () -> TacticT jdg ext err s m ()
+tweak f t = tactic $ \j -> mapExtract' f $ proofState t j
+
+-- | @peek t k@ lets us examine the extract produced by @t@, and then run a tactic based off it's value.
+peek :: (Functor m) => TacticT jdg ext err s m a -> (ext -> TacticT jdg ext err s m b) -> TacticT jdg ext err s m a
+peek t k = (tactic $ \j -> Subgoal ((), j) (\e -> fmap (first (const ())) $ proofState (k e) j)) >> t
+
+-- | @poke t k@ lets us examine the extract produced by @t@, and then run a tactic that produces a new extract.
+poke :: (Functor m) => TacticT jdg ext err s m () -> (ext -> TacticT jdg ext err s m ext) -> TacticT jdg ext err s m ()
```
Comment (isovector) [inline src/Refinery/Tactic.hs]: you absolute madman. you did it!

File: src/Refinery/Tactic.hs [hunk 5]
```diff
@@ -125,18 +135,44 @@ ensure p f t = check >> t
...
+-- Note that this function will backtrack on errors. If you want a version that provides partial proofs,
+-- use 'runPartialTacticT'
+runTacticT :: (MonadExtract ext m) => TacticT jdg ext err s m () -> jdg -> s -> m [Either err (Proof ext s jdg)]
 runTacticT t j s = proofs s $ fmap snd $ proofState t j
 
--- | Turn an inference rule into a tactic.
+-- | Runs a tactic, producing a list of possible extracts, along with a list of unsolved subgoals.
+-- Note that this function will produce a so called "Partial Proof". This means that we no longer backtrack on errors,
+-- but rather leave an unsolved hole, and continue synthesizing a extract.
+-- If you wnat a version that backgracks on errors, see 'runTacticT'.
```
Comment (isovector) [inline src/Refinery/Tactic.hs]:
```suggestion
-- If you want a version that backtracks on errors, see 'runTacticT'.
```

Comment (isovector) [reply src/Refinery/Tactic.hs]: Maybe a note about performance between this and `runTacticT`?

File: test/Spec.hs [hunk 6]
```diff
@@ -151,6 +152,7 @@ main = hspec $ do
     it "pure absorption on commit" $ property $ absorptionPureCommit (undefined :: TacticTest Int)
     it "empty identity on commit" $ property $ emptyIdentityCommit (undefined :: TacticTest Int)
     it "failure identity on commit" $ property $ emptyIdentityCommit (undefined :: TacticTest Int)
+    -- it "constant peek" $ property $ peekConst (undefined :: TacticTest ())
```
Comment (isovector) [inline test/Spec.hs]: Commented out on purpose?

File: src/Refinery/ProofState.hs (hunk 2 above)
Comment (TOTBWF) [reply src/Refinery/ProofState.hs]: Yes, will update the docs to reflect this

File: src/Refinery/Tactic.hs (hunk 4 above)
Comment (TOTBWF) [reply src/Refinery/Tactic.hs]: Immediately after the stream finished I realized how it would work lol

File: test/Spec.hs (hunk 6 above)
Comment (TOTBWF) [reply test/Spec.hs]: Yeah, test still fails for reasons that I don't totally understand

File: src/Refinery/ProofState.hs [hunk 7]
```diff
@@ -155,14 +189,56 @@ proofs s p = go s [] p
...
+      go _ _ (Failure err _) = pure [throwError err]
+      go s goals (Axiom ext) = pure [Right $ Proof ext s goals]
 
-accumEither :: (Semigroup a, Semigroup b) => Either a b -> Either a b -> Either a b
-accumEither (Left a1) (Left a2)   = Left (a1 <> a2)
-accumEither (Right b1) (Right b2) = Right (b1 <> b2)
-accumEither Left{} x              = x
-accumEither x Left{}              = x
+-- | The result of executing 'partialProofs'.
+data PartialProof ext s goal err
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: `ext` is a more natural type to be the last param.

Comment (isovector) [reply src/Refinery/ProofState.hs]: `PartialProof jdg err s ext` imo

File: src/Refinery/ProofState.hs [hunk 8]
```diff
@@ -155,14 +189,56 @@ proofs s p = go s [] p
...
+    = SuccessfulProof ext s [goal]
+    -- ^ A Proof that encountered no errors during the process of execution.
+    | PartialProof ext s [goal] [err]
+    -- ^ A so called "partial proof". These are proofs that encountered errors
+    -- during execution.
+    deriving (Eq, Show, Generic)
```
Comment (isovector) [inline src/Refinery/ProofState.hs:279]: +`Functor`

File: ChangeLog.md [hunk 9]
```diff
@@ -24,8 +24,10 @@
     ```
     data Proof ext s goal = Proof { pf_extract :: ext, pf_state :: s, pf_unsolvedGoals :: [goal] }
     ```
-  - `try` now uses `commit` instead of `<|>`. This means that we wont
-    generate a useless extract all the time.
+  - Removed `commit`, as it is fundementally broken.
```
Comment (isovector) [inline ChangeLog.md]: I think we want commit again after this morning :)

File: ChangeLog.md [hunk 10]
```diff
@@ -1,10 +1,37 @@
...
+  - A bunch of helpful combinators have been added for extract
+    manipulation inside of tactics.
+  - `proofs` no longer returns a tuple, but rather a record type
+    ```
+    data Proof ext s goal = Proof { pf_extract :: ext, pf_state :: s, pf_unsolvedGoals :: [goal] }
+    ```
+  - Added `handler`, and removed the `MonadError` instance for `TacticT`.
+    Now, instead of recovering from errors (which was fraught with subtle issues),
+	we allow the user to annotate errors instead.
+  - Added some useful tactic combinators
```
Comment (isovector) [inline ChangeLog.md]: Should list what they are

File: package.yaml [hunk 11]
```diff
@@ -22,28 +22,19 @@ description:         Please see the README on GitHub at <https://github.com/gith
...
 - base >= 4.7 && < 5
 - exceptions >= 0.10
-- logict >= 0.6
 - mmorph >= 1
 - mtl >= 2
 
 ghc-options:
   - -Wall
   - -Wredundant-constraints
+  - -fmax-relevant-binds=0
```
Comment (isovector) [inline package.yaml]: These might be better in your global config!

File: src/Refinery/ProofState.hs [hunk 12]
```diff
@@ -1,21 +1,23 @@
-{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE LambdaCase #-}
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: nit: formatting for these languages is all crazy

File: src/Refinery/ProofState.hs [hunk 13]
```diff
@@ -32,23 +34,48 @@ import           Control.Monad.Except
...
     = Subgoal goal (ext' -> ProofStateT ext' ext err s m goal)
+    -- ^ Represents a subgoal, and a continuation that tells
+    -- us what to do with the resulting extract.
     | Effect (m (ProofStateT ext' ext err s m goal))
+    -- ^ Run an effect.
     | Stateful (s -> (s, ProofStateT ext' ext err s m goal))
+    -- ^ Run a stateful computation. We don't want to use 'StateT' here, as it
+    -- doesn't play nice with backtracking.
     | Alt (ProofStateT ext' ext err s m goal) (ProofStateT ext' ext err s m goal)
+    -- ^ Combine together the results of two @ProofState@s.
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: Is there any reason to have alt?

File: src/Refinery/ProofState.hs [hunk 14]
```diff
@@ -126,93 +158,133 @@ instance (Monad m) => MonadPlus (ProofStateT ext ext err s m) where
...
+  default unsolvableHole :: (MonadTrans t, MonadExtract ext err m1, m ~ t m1) => err -> m ext
+  unsolvableHole = lift . unsolvableHole
+
+instance (MonadExtract ext err m) => MonadExtract ext err (ReaderT r m)
+instance (MonadExtract ext err m) => MonadExtract ext err (StateT s m)
+instance (MonadExtract ext err m, Monoid w) => MonadExtract ext err (LW.WriterT w m)
+instance (MonadExtract ext err m, Monoid w) => MonadExtract ext err (SW.WriterT w m)
+instance (MonadExtract ext err m) => MonadExtract ext err (ExceptT err m)
+
+data Proof ext s goal = Proof { pf_extract :: ext, pf_state :: s, pf_unsolvedGoals :: [goal] }
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: Some haddock here would be :+1:

File: src/Refinery/ProofState.hs [hunk 15]
```diff
@@ -126,93 +158,133 @@ instance (Monad m) => MonadPlus (ProofStateT ext ext err s m) where
...
+    deriving (Eq, Show, Generic)
+
+interleave :: [a] -> [a] -> [a]
+interleave (x : xs) (y : ys) = x : y : (interleave xs ys)
+interleave xs [] = xs
+interleave [] ys = ys
 
-proofs :: forall ext err s m goal. (MonadExtract ext m) => s -> ProofStateT ext ext err s m goal -> m [Either err (ext, s, [goal])]
-proofs s p = go s [] p
+prioritizing :: (forall a. [a] -> [a] -> [a])
```
Comment (isovector) [inline src/Refinery/ProofState.hs:226]: What's going on here?

File: src/Refinery/ProofState.hs [hunk 16]
```diff
@@ -126,93 +158,133 @@ instance (Monad m) => MonadPlus (ProofStateT ext ext err s m) where
...
+      go :: s -> [goal] -> (err -> m err) -> ProofStateT ext ext err s m goal -> m (Either [err] [Proof ext s goal])
+      go s goals handlers (Subgoal goal k) = do
+         h <- hole
+         (go s (goals ++ [goal]) handlers $ k h)
+      go s goals handlers (Effect m) = m >>= go s goals handlers
+      go s goals handlers (Stateful f) =
+          let (s', p) = f s
+          in go s' goals handlers p
+      go s goals handlers (Alt p1 p2) = liftA2 (prioritizing (<>)) (go s goals handlers p1) (go s goals handlers p2)
+      go s goals handlers (Interleave p1 p2) = liftA2 (prioritizing interleave) (go s goals handlers p1) (go s goals handlers p2)
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: nit: newline

File: test/Spec/PropertyTests.hs [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+    testBatch $ functor     (undefined :: TacticTest ((), (), ()))
+    testBatch $ applicative (undefined :: TacticTest ((), (), ()))
+    testBatch $ alternative (undefined :: TacticTest ())
+    testBatch $ monad       (undefined :: TacticTest ((), (), ()))
+    testBatch $ monadPlus   (undefined :: TacticTest ((), ()))
+    testBatch $ monadState  (undefined :: TacticTest ((), ()))
+    it "interleave - mzero" $ property $ interleaveMZero (undefined :: TacticTest Int)
+    it "interleave - mplus" $ property $ interleaveMPlus (undefined :: TacticTest Int)
+    it "distrib put over <|>" $ property $ distribPut (undefined :: TacticTest ())
+    -- it "constant peek" $ property $ peekConst (undefined :: TacticTest ())
```
Comment (isovector) [inline test/Spec/PropertyTests.hs:167]: Tests for `handling`?

File: src/Refinery/ProofState.hs [hunk 18]
```diff
@@ -126,93 +158,133 @@ instance (Monad m) => MonadPlus (ProofStateT ext ext err s m) where
...
+          _                              -> go s goals handlers p2
+      go _ _ _ Empty = pure $ Left []
+      go _ _ handlers (Failure err _) = do
+          annErr <- handlers err
+          pure $ Left [annErr]
+      go s goals handlers (Handle p h) =
+          -- NOTE [Handler ordering]:
+          -- If we have multiple handlerss in scope, then we want the ones further down the stack to
+          -- run /first/. This allows the handlerss up the stack to add their annotations on top of the
+          -- ones lower down, which is the behavior that we desire.
```
Comment (isovector) [inline src/Refinery/ProofState.hs:269]: This is not self-evident to me. Why is this preferable?

File: src/Refinery/ProofState.hs (hunk 13 above)
Comment (TOTBWF) [reply src/Refinery/ProofState.hs]: Yeah, it has to do with interleaving of results + the identity laws.

File: src/Refinery/ProofState.hs [hunk 19]
```diff
@@ -126,93 +158,133 @@ instance (Monad m) => MonadPlus (ProofStateT ext ext err s m) where
...
 instance (Monad m) => MonadState s (ProofStateT ext ext err s m) where
     state f = Stateful $ \s ->
       let (a, s') = f s
       in (s', pure a)
 
+{-# DEPRECATED axiom "Use Axiom instead" #-}
 axiom :: ext -> ProofStateT ext' ext err s m jdg
 axiom = Axiom
 
+-- | @subgoals fs p@ will apply a list of functions producing a new 'ProofState' to each of the subgoals of @p@.
```
Comment (isovector) [inline src/Refinery/ProofState.hs]:
```suggestion
-- | @subgoals fs p@ will apply a list of functions producing a new 'ProofState' to each of the subgoals of @p@ in order.
```

File: src/Refinery/ProofState.hs [hunk 20]
```diff
@@ -221,30 +293,22 @@ subgoals fs (Stateful s) = Stateful (fmap (subgoals fs) . s)
...
-    Alt t1 t2 -> Alt (mapExtract' into t1) (mapExtract' into t2)
-    Interleave t1 t2 -> Interleave (mapExtract' into t1) (mapExtract' into t2)
-    Commit t1 t2 -> Commit (mapExtract' into t1) (mapExtract' into t2)
-    Empty -> Empty
-    Failure err -> Failure err
-    Axiom ext -> Axiom $ into ext
+-- | @mapExtract f g p@ allows yout to modify the extract type of a ProofState.
+-- This witness the @Profunctor@ instance of 'ProofState', which we can't write without a newtype due to
+-- the position of the type variables
+mapExtract :: (Functor m) => (a -> ext') -> (ext -> b) -> ProofStateT ext' ext err s m jdg -> ProofStateT a b err s m jdg
```
Comment (isovector) [inline src/Refinery/ProofState.hs:343]: Nice simplification!

File: src/Refinery/Tactic.hs [hunk 21]
```diff
@@ -62,27 +80,67 @@ infixr 3 <%>
...
 try t = t <|> pure ()
 
--- | Runs a tactic repeatedly until it fails
+-- | @commit t1 t2@ will run @t1@, and then run @t2@ only if @t1@ failed to produce any successes.
+--
+-- NOTE: @commit@ does have some sneaky semantics that you have to be aware of. Specifically, it interacts a bit
+-- surprisingly with '>>='. Namely, the following algebraic law holds
+-- @
+--     commit t1 t2 >>= f = commit (t1 >>= f) (t2 >>= f)
+-- @
```
Comment (isovector) [inline src/Refinery/Tactic.hs:100]: A note here about why this leads to surprises would be helpful

File: src/Refinery/Tactic.hs [hunk 22]
```diff
@@ -62,27 +80,67 @@ infixr 3 <%>
...
+commit :: TacticT jdg ext err s m a -> TacticT jdg ext err s m a -> TacticT jdg ext err s m a
+commit t1 t2 = tactic $ \j -> Commit (proofState t1 j) (proofState t2 j)
+
+-- | Runs a tactic 0 or more times until it fails.
+-- Note that 'many_' is almost always the right choice over 'many'.
+-- Due to the semantics of 'Alternative', 'many' will create a bunch
+-- of partially solved goals along the way, one for each iteration.
```
Comment (isovector) [inline src/Refinery/Tactic.hs:110]: I think you could get away with an `Applicative` constraint instead if you used `*>`

File: src/Refinery/Tactic.hs [hunk 23]
```diff
@@ -62,27 +80,67 @@ infixr 3 <%>
...
 many_ :: (Monad m) => TacticT jdg ext err s m () -> TacticT jdg ext err s m ()
 many_ t = try (t >> many_ t)
 
+-- | Runs a tactic 1 or more times until it fails.
+-- Note that 'some_' is almost always the right choice over 'some'.
+-- Due to the semantics of 'Alternative', 'some' will create a bunch
+-- of partially solved goals along the way, one for each iteration.
+some_ :: (Monad m) => TacticT jdg ext err s m () -> TacticT jdg ext err s m ()
+some_ t = t >> many_ t
```
Comment (isovector) [inline src/Refinery/Tactic.hs:119]: same

File: src/Refinery/Tactic.hs [hunk 24]
```diff
@@ -92,27 +150,27 @@ progress eq err t = do
...
 -- the next tactic to run.
-gather :: (MonadExtract ext m) => TacticT jdg ext err s m a -> ([(a, jdg)] -> TacticT jdg ext err s m a) -> TacticT jdg ext err s m a
+gather :: (MonadExtract ext err m) => TacticT jdg ext err s m a -> ([(a, jdg)] -> TacticT jdg ext err s m a) -> TacticT jdg ext err s m a
 gather t f = tactic $ \j -> do
     s <- get
     results <- lift $ proofs s $ proofState t j
-    msum $ flip fmap results $ \case
-        Left err -> throwError err
-        Right (_, _, jdgs) -> proofState (f jdgs) j
+    case results of
```
Comment (isovector) [inline src/Refinery/Tactic.hs]: What does this function do for partial proofs? Might be worth mentioning in the haddock

File: src/Refinery/Tactic.hs [hunk 25]
```diff
@@ -92,27 +150,27 @@ progress eq err t = do
...
+      Left errs -> msum $ fmap (\err -> Failure err Axiom) errs
```
Comment (isovector) [inline src/Refinery/Tactic.hs]: is `msum . fmap` just `foldMap`?

File: src/Refinery/Tactic.hs [hunk 26]
```diff
@@ -92,27 +150,27 @@ progress eq err t = do
...
+      Right pfs -> msum $ fmap (\(Proof _ _ jdgs) -> proofState (f jdgs) j) pfs
```
Comment (isovector) [inline src/Refinery/Tactic.hs]: nit: you'll appreciate it in the future when `Proof` changes if you change this pattern match to be a field selector

File: src/Refinery/Tactic.hs [hunk 27]
```diff
@@ -125,18 +183,51 @@ ensure p f t = check >> t
           s <- get
           modify f
           case p s of
-            Just err -> throwError err
+            Just err -> unsolvable err
```
Comment (isovector) [inline src/Refinery/Tactic.hs:178]: Is there a difference between `unsolvable` and `unsolvableHole` and `failure`?

File: src/Refinery/Tactic.hs [hunk 28]
```diff
@@ -125,18 +183,51 @@ ensure p f t = check >> t
...
-runTacticT :: (MonadExtract ext m) => TacticT jdg ext err s m () -> jdg -> s -> m [Either err (ext, s, [jdg])]
+-- Note that this function will backtrack on errors. If you want a version that provides partial proofs,
+-- use 'runPartialTacticT'
+runTacticT :: (MonadExtract ext err m) => TacticT jdg ext err s m () -> jdg -> s -> m (Either [err] [(Proof ext s jdg)])
 runTacticT t j s = proofs s $ fmap snd $ proofState t j
 
--- | Turn an inference rule into a tactic.
+-- | Run a tactic, and get just the list of extracts, ignoring any other information.
+solutions :: (MonadExtract ext err m) => TacticT jdg ext err s m () -> jdg -> s -> m [ext]
+solutions t j s = either (const []) (map pf_extract) <$> runTacticT t j s
```
Comment (isovector) [inline src/Refinery/Tactic.hs]: `evalTacticT` might be a better name here

File: src/Refinery/Tactic/Internal.hs [hunk 29]
```diff
@@ -57,17 +58,18 @@ import Refinery.ProofState
 -- * @s@   - The state type.
 -- * @m@   - The base monad.
 -- * @a@   - The return value. This to make @'TacticT'@ a monad, and will always be @'()'@
+--
+-- One of the most important things about this type is it's 'Monad' instance. @t1 >> t2@
```
Comment (isovector) [inline src/Refinery/Tactic/Internal.hs:63]: I love this comment. Might  be worth spelling out "THIS IS DIFFERENT THAN EVERY OTHER MONAD YOU'VE EVER PLAYED WITH"

File: stack-hie.yaml [hunk 30]
```diff
@@ -0,0 +1,67 @@
+# This file was automatically generated by 'stack init'
```
Comment (isovector) [inline stack-hie.yaml:1]: Is this file necessary? It broke my hls setup when I tried to fire up refinery.

File: ChangeLog.md (hunk 9 above)
Comment (TOTBWF) [reply ChangeLog.md]: It has RETURNED

File: src/Refinery/Tactic.hs (hunk 22 above)
Comment (TOTBWF) [reply src/Refinery/Tactic.hs:110]: Sadly not, due to that `Effect` ctor

File: stack-hie.yaml (hunk 30 above)
Comment (TOTBWF) [reply stack-hie.yaml:1]: It prevents HLS from trashing builds/corrupting stuff, which it can do from time to time.

File: src/Refinery/Tactic.hs (hunk 27 above)
Comment (TOTBWF) [reply src/Refinery/Tactic.hs:178]: Yeah, it's a matter of context.

File: src/Refinery/ProofState.hs (hunk 12 above)
Comment (TOTBWF) [reply src/Refinery/ProofState.hs]: Yeah HLS likes to do weird stuff sometimes

File: src/Refinery/Tactic.hs (hunk 24 above)
Comment (TOTBWF) [reply src/Refinery/Tactic.hs]: Going to see if I can do this "right"

File: src/Refinery/ProofState.hs [hunk 31]
```diff
@@ -167,14 +169,32 @@ instance (MonadExtract ext err m, Monoid w) => MonadExtract ext err (LW.WriterT
 instance (MonadExtract ext err m, Monoid w) => MonadExtract ext err (SW.WriterT w m)
 instance (MonadExtract ext err m) => MonadExtract ext err (ExceptT err m)
 
-data Proof s goal ext = Proof { pf_extract :: ext, pf_state :: s, pf_unsolvedGoals :: [goal] }
+-- | Represents a single result of the execution of some tactic.
+data Proof s goal ext = Proof
```
Comment (isovector) [inline src/Refinery/ProofState.hs]: It'd be nice to have a functor instance for this!

File: test/Spec.hs (hunk 6 above)
Comment (TOTBWF) [reply test/Spec.hs]: Ok they fail because `peek` is broken who woulda thunk. reify machinery does this right.

Comment (isovector) [reply test/Spec.hs]: wow, tests hey?

**Comment by isovector** [comment]
The `handler` appears to leak into the "upwards" continuations (in the official TOTBWF refinery diagrams.)

stlc repro:

```haskell
bug :: Either [String] [Term]
bug = evalT bug' ([("a", "a")] :- TPair "a" "b") 0
  where
    bug' = do
      pair
      choice
        [ do
            _ :- g <- goal
            handler $ \err -> do
              traceM $ "handling " <> show g <> " while " <> err
              pure err
        , assumption
        , do
            _ :- g <- goal
            failure $ "erroring on " <> show g
        ]
```

whose output is:

```
handling TVar "a" while goal mismatch: Assumption
handling TVar "a" while erroring on TVar "b"
```

The second line here is concerning. It's the handler that we installed in the first hole (indicated by the goal), leaking into the second (indicated by the error.)

**Comment by isovector** [comment]
After playing with it for a while, it seems like the usual type for `catchError` (rather than `handler`) solves the scoping issues here.
